tests/gtests/internals/test_sdpa.cpp

```diff
@@ -0,0 +1,1916 @@
+    }
+    if (!(p.kdt == mdt::f16 || p.vdt == mdt::f16)
+            && !(p.kdt == mdt::bf16 || p.vdt == mdt::f16)) {
+        ss << "_qtype_" << p.qtype;
+    }
+    return ss;
+}
+
+std::string PrintToString(const ::testing::TestParamInfo<sdpa_dims_t> &info) {
+    std::stringstream ss;
+    ss << info.param;
+    return ss.str();
+}
+
+void print_test_case(memory::data_type dt, const sdpa_dims_t &p) {
+    std::cout << '[' << std::setw(4) << dnnl_dt2str(memory::convert_to_c(dt));
+    std::cout << " mb(n) = " << p.mb << ", seq_len(k) = " << p.seq_len
+              << ", head_num = " << p.head_num
+              << ", head_size(d/v) = " << p.head_size
+              << ", query_num(q) = " << p.query_num
+              << ", kgroup_size = " << p.kgroup_size
+              << ", vgroup_size = " << p.vgroup_size;
+    std::cout << "] " << std::flush;
+}
+
+std::string print_table_header(const sdpa_dims_t &p) {
+    std::stringstream ss;
+    ss << "| mb |    K | #Head |   D |    q | Kdt | Vdt |  time |";
+    return ss.str();
+}
+
+std::string print_row(const sdpa_dims_t &p) {
+    std::stringstream ss;
+
+    ss << "|" << p.mb;
+    ss << "|" << p.seq_len;
+    ss << "|" << p.head_num;
+    ss << "|" << p.head_size;
+    ss << "|" << p.query_num;
+    ss << "|" << p.kdt;
+    if (p.kdt != mdt::f16 && p.qtype != quantize_type::no_quantization) {
+        ss << "|" << p.ksdt;
+        ss << "|" << p.kzpdt;
+    }
+    ss << "|" << p.vdt;
+    if (p.vdt != mdt::f16 && p.qtype != quantize_type::no_quantization) {
+        ss << "|" << p.vsdt;
+        ss << "|" << p.vzpdt;
+    }
+    switch (p.mask) {
+        case mask_type::no_mask: ss << "_no_mask"; break;
+        case mask_type::oneD: ss << "_mask1D"; break;
+        case mask_type::twoD: ss << "_mask2D"; break;
+        case mask_type::causal: ss << "_maskcausal"; break;
+    }
+    if (p.kdt != mdt::f16 || p.vdt != mdt::f16) { ss << "|" << p.qtype; }
+    return ss.str();
+}
+
+using dnnl::algorithm;
+using dnnl::matmul;
+using dnnl::memory;
+using dnnl::primitive_attr;
+using dnnl::softmax_forward;
+
+#define COMPLAIN_DNNL_ERROR_AND_EXIT(what, status) \
+    do { \
+        printf("[%s:%d] `%s` returns oneDNN error: %s.\n", __FILE__, __LINE__, \
+                what, dnnl_status2str(status)); \
+        printf("Example failed.\n"); \
+        exit(1); \
+    } while (0)
+
+#define COMPLAIN_EXAMPLE_ERROR_AND_EXIT(complain_fmt, ...) \
+    do { \
+        printf("[%s:%d] Error in the example: " complain_fmt ".\n", __FILE__, \
+                __LINE__, __VA_ARGS__); \
+        printf("Example failed.\n"); \
+        exit(2); \
+    } while (0)
+
+#undef CHECK
+#define CHECK(f) \
+    do { \
+        dnnl_status_t s_ = f; \
+        if (s_ != dnnl_success) COMPLAIN_DNNL_ERROR_AND_EXIT(#f, s_); \
+    } while (0)
+
+// Read from handle, write to memory
+template <typename T>
+inline void write_to_dnnl_memory(const T *handle, dnnl::memory &mem) {
```

Comment:
Minor: have you checked if including examples_utils.hpp would help to reduce the amount of the code in this test?

Reply:
That version of the write_to_dnnl_memory doesn't handle the data types correctly. This is more flexible wrt the source and destination data type.

Reply:
Would you consider to enhance that version and re-use it?

Reply:
I feel like this version is being held together with duct tape and its more complicated to read compared to the one in the examples file. Personally I feel the example functions should be easy enough to read for new users and should be as simple as possible.